### PR TITLE
Add support for Arrow IPC Stream files

### PR DESCRIPTION
**Pull request by joshuataylor**
This is waiting on https://github.com/pola-rs/polars/pull/3783 being merged, once it is we can remove the reference to my repository, and just use the Polars version with this.

- [x] Merged https://github.com/jorgecarleitao/arrow2/pull/1097
- [x] Merged https://github.com/pola-rs/polars/pull/3793
- [x] Merged https://github.com/pola-rs/polars/pull/3783
- [x] Minor version bump of Polars on https://crates.io/crates/polars/versions

Closes https://github.com/elixir-nx/explorer/issues/265

### CODE REVIEW

**Comment by joshuataylor** [comment]
This is now merged in Polars, so we wait for this to be released in Polars. 🎉

**Comment by josevalim** [comment]
Congratulations on shipping this upstream!

**Comment by cigrainger** [comment]
Congrats @joshuataylor! Great stuff. 🎉

**Comment by josevalim** [comment]
@joshuataylor can you please rebase? I think you can pretty much discard your changes to `expressions.rs` and we should be good to go!

**Comment by joshuataylor** [comment]
:eyes:

File: lib/explorer/backend/data_frame.ex [hunk 1]
```diff
@@ -56,6 +56,14 @@ defmodule Explorer.Backend.DataFrame do
   @callback to_ipc(df, filename :: String.t(), compression :: {nil | atom(), nil | integer()}) ::
               ok_result()
 
+
```
Comment (josevalim) [inline lib/explorer/backend/data_frame.ex:59]:
```suggestion
```

File: lib/explorer/data_frame.ex [hunk 2]
```diff
@@ -544,6 +545,69 @@ defmodule Explorer.DataFrame do
...
+    * `columns` - List with the name or index of columns to be selected. Defaults to all columns.
+  """
+  @doc type: :io
+  @spec from_ipc_stream(filename :: String.t()) :: {:ok, DataFrame.t()} | {:error, term()}
+  def from_ipc_stream(filename, opts \\ []) do
+    opts =
+      Keyword.validate!(opts,
+        columns: nil
+      )
+
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:560]:
```suggestion
    opts = Keyword.validate!(opts, columns: nil)
```

File: lib/explorer/data_frame.ex [hunk 3]
```diff
@@ -544,6 +545,69 @@ defmodule Explorer.DataFrame do
...
+    backend = backend_from_options!(opts)
+
+    backend.from_ipc_stream(
+      filename,
+      opts[:columns]
+    )
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:566]:
```suggestion
    backend.from_ipc_stream(filename, opts[:columns])
```

File: lib/explorer/data_frame.ex [hunk 4]
```diff
@@ -544,6 +545,69 @@ defmodule Explorer.DataFrame do
     backend.to_ipc(df, filename, {compression, nil})
   end
 
+  @doc """
+  Reads an IPC Streaming file into a dataframe.
+  ## Options
+    * `columns` - List with the name or index of columns to be selected. Defaults to all columns.
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:551]:
```suggestion
  Reads an IPC Streaming file into a dataframe.

  ## Options

    * `columns` - List with the name or index of columns to be selected. Defaults to all columns.
```

File: lib/explorer/data_frame.ex [hunk 5]
```diff
@@ -544,6 +545,69 @@ defmodule Explorer.DataFrame do
...
+        compression: nil
+      )
+
+    backend = backend_from_options!(opts)
+
+    backend.to_ipc_stream(
+      df,
+      filename,
+      opts[:compression]
+    )
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:608]:
```suggestion
    opts = Keyword.validate!(opts, compression: nil)
    backend = backend_from_options!(opts)

    backend.to_ipc_stream(df, filename, opts[:compression])
```

File: lib/explorer/data_frame.ex [hunk 6]
```diff
@@ -544,6 +545,69 @@ defmodule Explorer.DataFrame do
...
+  Writes a dataframe to a IPC Stream file.
+  Arrow IPC Streams provide a streaming protocol or “format" for sending an arbitrary length sequence of record batches.
+  The format must be processed from start to end, and does not support random access.
+  You can read more information about the difference between IPC and IPC Streaming files on the
+  [https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format](Apache Arrow Documentation(#ipc-streaming-format) page.
+  If you want to write a dataframe IPC, it's recommended to use IPC instead, unless you really need to as using IPC allows
+  supports random access, and thus is very useful when used with memory maps.
+  ## Options
+    * `compression` - Sets the algorithm used to compress the IPC file.
+      It accepts `"ZSTD"` or `"LZ4"` compression. (default: `nil`)
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:591]:
```suggestion
  Writes a dataframe to a IPC Stream file.

  Arrow IPC Streams provide a streaming protocol or “format" for sending an arbitrary length sequence of record batches.
  The format must be processed from start to end, and does not support random access.

  You can read more information about the difference between IPC and IPC Streaming files on the
  [https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format](Apache Arrow Documentation(#ipc-streaming-format) page.

  ## Options

    * `compression` - Sets the algorithm used to compress the IPC file.
      It accepts `"ZSTD"` or `"LZ4"` compression. (default: `nil`)
```

File: lib/explorer/polars_backend/data_frame.ex [hunk 7]
```diff
@@ -150,6 +150,24 @@ defmodule Explorer.PolarsBackend.DataFrame do
...
+    case Native.df_read_ipc_stream(filename, columns, projection) do
+      {:ok, df} -> {:ok, Shared.create_dataframe(df)}
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  @impl true
+  def to_ipc_stream(%DataFrame{data: df}, filename, compression) do
+    case Native.df_write_ipc_stream(df, filename, compression) do
+      {:ok, _} -> {:ok, filename}
```
Comment (josevalim) [inline lib/explorer/polars_backend/data_frame.ex:166]:
```suggestion
      {:ok, _} -> :ok
```

File: lib/explorer/data_frame.ex [hunk 8]
```diff
@@ -544,6 +545,58 @@ defmodule Explorer.DataFrame do
...
+  [https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format](Apache Arrow Documentation(#ipc-streaming-format) page.
+
+  ## Options
+
+    * `compression` - Sets the algorithm used to compress the IPC file.
+      It accepts `"ZSTD"` or `"LZ4"` compression. (default: `nil`)
+  """
+  @doc type: :io
+  @spec to_ipc_stream(df :: DataFrame.t(), filename :: String.t()) ::
+          {:ok, String.t()} | {:error, term()}
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:592]:
```suggestion
          :ok | {:error, term()}
```

**Review by philss — APPROVED** [review]
Awesome work! :rocket:

**Pull request merged**